core/trino-main/src/main/java/io/trino/operator/aggregation/listagg/ListaggAggregationFunction.java

```diff
@@ -0,0 +1,335 @@
+        }
+
+        state.merge(otherState);
+    }
+
+    public static void output(Type type, ListaggAggregationState state, BlockBuilder out)
+    {
+        if (state.isEmpty()) {
+            out.appendNull();
+        }
+        else {
+            outputState(state, out, MAX_OUTPUT_LENGTH);
+        }
+    }
+
+    @VisibleForTesting
+    protected static void outputState(ListaggAggregationState state, BlockBuilder out, int maxOutputLength)
+    {
+        Slice separator = state.getSeparator();
+        AtomicLong aggregationLength = new AtomicLong(0L);
+        AtomicInteger nonNullValuesCount = new AtomicInteger(0);
+        state.forEach((block, position) -> {
+            if (!block.isNull(position)) {
+                if (aggregationLength.get() > 0) {
+                    aggregationLength.addAndGet(separator.length());
+                }
+                aggregationLength.addAndGet(block.getSliceLength(position));
+                nonNullValuesCount.incrementAndGet();
+            }
+        });
+        if (nonNullValuesCount.get() == 0) {
+            out.appendNull();
+        }
+        else {
+            if (aggregationLength.get() > maxOutputLength) {
+                if (state.isOverflowError()) {
+                    throw new TrinoException(QUERY_REJECTED, "Concatenated string is too large");
+                }
+                else {
+                    outputTruncatedState(state, out, nonNullValuesCount.get(), maxOutputLength);
+                }
+            }
+            else {
+                AtomicBoolean elementWritten = new AtomicBoolean(false);
+                state.forEach((block, position) -> {
+                    if (!block.isNull(position)) {
+                        if (elementWritten.get()) {
+                            out.writeBytes(separator, 0, separator.length());
+                        }
+                        block.writeBytesTo(position, 0, block.getSliceLength(position), out);
+                        if (!elementWritten.get()) {
+                            elementWritten.set(true);
+                        }
+                    }
+                });
+            }
+            out.closeEntry();
+        }
+    }
+
+    private static void outputTruncatedState(ListaggAggregationState state, BlockBuilder out, int nonNullValuesCount, int maxOutputLength)
+    {
+        Slice separator = state.getSeparator();
+        Slice overflowTruncationFiller = state.getOverflowTruncationFiller();
+        AtomicInteger truncatedAggregationLength = new AtomicInteger(0);
+        AtomicInteger nonNullValuesWritten = new AtomicInteger(0);
+
+        int overflowTruncationFillerLength = overflowTruncationFiller.length();
+        int maxOutputLengthWithoutTruncationFilling = state.isOverflowTruncationCountIndication() ?
+                maxOutputLength - overflowTruncationFillerLength - 2 /*()*/ - nonNullValuesCount :
```

Comment:
The `/*()*/` comment is very cryptic. I assume it's to indicate that 2 is the length of `()`. I would spell it out more clearly: `/* length of () */

Also, I don't understand how `nonNullValuesCount` fits in this formula. It's doing math between lengths, but nonNullValuesCount is a number of elements, not a length.

Reply:
> Also, I don't understand how nonNullValuesCount fits in this formula. It's doing math between lengths, but nonNullValuesCount is a number of elements, not a length.

I don't remember exactly what `WITH COUNT` is supposed to do because unfortunately I don't have the SQL standard document at hand.

I remember that I got inspired in this regard from:

https://modern-sql.com/feature/listagg

> If with count is specified and truncation happens, the number of omitted values is put in brackets and appended to the result.

That's why I thought that `nonNullValuesCount` fits within the formula.

Reply:
Right, that's how it's supposed to work, if I recall correctly. But in that case you'd need to know the length of the rendered string corresponding to the number of values, not just add the number of values directly.

Reply:
Sorry, I've interpreted falsely initially this aspect.

I think that the overflow count indication should be ok now